tests/hooks/use-fetch.test.js

```diff
@@ -0,0 +1,83 @@
+import useFetch from '../../src/hooks/use-fetch';
+import { cleanup, renderHook } from '@testing-library/react-hooks';
+
+const sampleSuccess = {
+    someResult:
+        "I'm an example of a success response. This hook is meant to provide #request() capabilities in hook format",
+};
+
+const mockSuccessApi = {
+    ok: true,
+    json() {
+        return sampleSuccess;
+    },
+};
+
+const mockFailApi = {
+    ok: false,
+    json() {
+        return new Error('An error occured');
+    },
+};
+
+describe('Use Fetch', () => {
+    beforeEach(() => {
+        window.fetch = jest.fn();
+    });
+
+    test('should get data', async () => {
+        window.fetch.mockReturnValue(mockSuccessApi);
+        const { result, waitForNextUpdate } = renderHook(() =>
+            useFetch('/url', x => x.json())
+        );
+        await waitForNextUpdate();
+        const { data, error } = result.current;
+        expect(data).toEqual(sampleSuccess);
+        expect(error).toEqual(null);
+        cleanup();
+    });
+
+    test('should return an error if any are thrown', async () => {
+        const { warn } = console;
+        console.warn = jest.fn();
+        window.fetch.mockReturnValue(mockFailApi);
+        const { result, waitForNextUpdate } = renderHook(() =>
+            useFetch('/url', x => x.json())
+        );
+        await waitForNextUpdate();
+        const { data, error } = result.current;
+        expect(data).toEqual(null);
+        expect(console.warn).toHaveBeenCalledTimes(1);
+        expect(error).not.toBeNull();
+        console.warn = warn;
+        cleanup();
+    });
+
+    test('should debounce fetch if requested', async () => {
+        // Use fake timers for easier control over debounce
+        jest.useFakeTimers();
+        window.fetch.mockReturnValue(mockSuccessApi);
+        let url = '/url';
+        const { result, rerender, waitForNextUpdate } = renderHook(() =>
+            useFetch(url, x => x.json(), 1000)
+        );
+        expect(window.fetch).toHaveBeenCalledTimes(0);
+        let { data, error } = result.current;
+        expect(data).toEqual(null);
+        expect(error).toEqual(null);
+        // Update the hook 10 additional times before running timers,
+        // only one call should occur (last one)
+        for (let i = 0; i < 10; i++) {
```

Comment:
Interesting, how does this help with testing the debounce functionality?

Reply:
Ah so what this does is the `rerender` function will call a new `useEffect` each time (since the url has been updated in the loop). That `useEffect` then sets the timeout to actually fetch the data. With jest we can `useFakeTimers` so this batches the events up so we don't have timers going off until we want. Together (if I did this right) we can keep calling the internal `useEffect` which keeps setting (and clearing) timeouts and so when we run the timers we only hit `fetch` once with the last url.

Reply:
That definitely makes sense, really like this approach 👍